File: snake.js
function adjacentValidSquare(square) {
  return {
    x: square.x,
    y: square.y - 1,
  }
}

function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min) + min);
}

const UP = 0, RIGHT = 1, DOWN = 2, LEFT = 3;
const SCALE = 20;

function Snake(snakeGame, snakeNet, snakeAI, player) {
  this.size = 5;
  this.currentSpaces = [];
  this.currentDirection = 2;
  this.growAmount = 0;
  lastKeyCode = 0; //down
  
  for(let i = 0;i < this.size;i ++) {
    if(i == 0) {
      this.currentSpaces.push({
        x: randomInt(0, snakeGame.gridWidth),
        y: randomInt(0, snakeGame.gridHeight - this.size)
      });
    } else {
      this.currentSpaces.push(adjacentValidSquare(this.currentSpaces[i - 1]));
    }
  }
  
  this.getNeckDirection = function() {
    if(this.currentSpaces[0].x < this.currentSpaces[1].x) return RIGHT;
    if(this.currentSpaces[0].y > this.currentSpaces[1].y) return UP;
    if(this.currentSpaces[0].y < this.currentSpaces[1].y) return DOWN;
    if(this.currentSpaces[0].x > this.currentSpaces[1].x) return LEFT;
    throw "Bad neck " + this.currentSpaces;
  }
  
  this.selectDirection = function() {
    /*let dirs = [0, 1, 2, 3];
    dirs.splice(dirs.indexOf(this.getNeckDirection()), 1);
    return dirs[randomInt(0,3)];*/
    let ret = this.currentDirection;
    if(lastKeyCode == 38) ret = UP;
    if(lastKeyCode == 39) ret = RIGHT;
    if(lastKeyCode == 40) ret = DOWN;
    if(lastKeyCode == 37) ret = LEFT;
    return ret;
  }
  
  let dir;
  this.step1 = function(state, features) {
    let ndir = this.selectDirectionNet(features);
    if(player == HUMAN) {
      dir = this.selectDirection();
    } else if(player == NEURAL) {
      dir = ndir;
    } else if(player == AI) {
      dir = this.selectDirectionAI(state);
    }
    this.currentDirection = dir;
    return [getNewSpace(this.currentSpaces[0], dir), getNewSpace(this.currentSpaces[0], ndir)];
  }

  this.step2 = function() {
    if(!this.growAmount) {
      let temp = getNewSpace(this.currentSpaces[0], dir);
      for(let i = this.currentSpaces.length - 2;i >= 0;i --) {
        this.currentSpaces[i + 1] = this.currentSpaces[i];
      }
      this.currentSpaces[0] = temp;
    } else {
      this.growAmount--;
      this.size++;
      for(let i = this.currentSpaces.length - 1;i >= 0;i --) {
        this.currentSpaces[i + 1] = this.currentSpaces[i];
      }
      this.currentSpaces[0] = getNewSpace(this.currentSpaces[0], dir);
    }
  }
  
  this.isDead = function() {
    let dead = false;
    let x = this.currentSpaces[0].x;
    let y = this.currentSpaces[0].y;
    if(x < 0 || x >= snakeGame.gridWidth || y < 0 || y >= snakeGame.gridHeight) return true;
    this.currentSpaces.forEach((square, index) => {
      if(!index || dead) return;
      if(x == square.x && y == square.y) dead = true;
    });
    return dead;
  }
  
  this.hitFood = function(food) {
    return food.x == this.currentSpaces[0].x && food.y == this.currentSpaces[0].y;
  }
  
  this.grow = function(amount) {
    this.growAmount += amount;
  }
  
  //leftblocked, frontblocked, rightblocked, dh(norm), dv(norm);
  
  this.selectDirectionNet = function(features) {
    let dir = snakeNet.selectDirection(features);
    let ret = (this.currentDirection + dir + 3) % 4;
    return ret;
  }

  this.selectDirectionAI = function(state) {
    let spaces = expandSpaces(state.s);
    let dir = snakeAI.selectDirection(spaces, state.f);
    let ret = (this.currentDirection + dir + 3) % 4;
    return ret;
  }
}

let lastKeyCode = 0;
window.onkeydown = function(e) { lastKeyCode = e.keyCode; }

File: python/snake.py
import pygame
from pygame.locals import K_UP, K_RIGHT, K_DOWN, K_LEFT
from util import Space, Player, getNewSpace, printList, expandSpaces
from random import randint

UP = 0
RIGHT = 1
DOWN = 2
LEFT = 3

class Snake:
    def __init__(self, snakeGame, snakeNet, snakeAI, player):
        self.size = 5
        self.direction = DOWN
        self.spaces = []
        self.grow = 0
        self.snakeGame = snakeGame
        self.stepDir = 0
        self.player = player
        self.snakeNet = snakeNet
        self.snakeAI = snakeAI
        for i in range(0, self.size):
            if i == 0:
                self.spaces.append(Space(
                        randint(0, snakeGame.gridWidth - 1),
                        randint(3, snakeGame.gridHeight - self.size - 3)
                    ))
            else:
                self.spaces.append(self.adjacentSquare(self.spaces[i - 1]))
    
    def step1(self, state, features):
        #ndir = self.selectDirectionNet(features)
        if self.player == Player.HUMAN:
            self.stepDir = self.selectDirection()
        #elif self.player == Player.NEURAL:
            #self.stepDir = ndir
        elif self.player == Player.AI:
            self.stepDir = self.selectDirectionAI(state)
        self.direction = self.stepDir
        return (getNewSpace(self.spaces[0], self.direction),) #getNewSpace(self.spaces[0], ndir))

    def step2(self):
        if self.grow > 0:
            self.grow -= 1
            self.size += 1
            self.spaces.append(None)
            for i in range(len(self.spaces) - 2, -1, -1):
                self.spaces[i + 1] = self.spaces[i]
            self.spaces[0] = getNewSpace(self.spaces[0], self.direction)
        else:
            newSpace = getNewSpace(self.spaces[0], self.direction)
            for i in range(len(self.spaces) - 2, -1, -1):
                self.spaces[i + 1] = self.spaces[i]
            self.spaces[0] = newSpace

    def adjacentSquare(self, space):
        return Space(space.x, space.y - 1)

    def selectDirection(self):
        dir = self.direction
        pygame.event.pump()
        keys = pygame.key.get_pressed()
        if keys[K_LEFT]: dir = LEFT
        if keys[K_UP]: dir = UP
        if keys[K_DOWN]: dir = DOWN
        if keys[K_RIGHT]: dir = RIGHT
        return dir
    
    def selectDirectionNet(self, features):
        turn = self.snakeNet.selectDirection(features)
        dir = (self.direction + turn + 3) % 4
        return dir
    
    def selectDirectionAI(self, state):
        spaces = expandSpaces(state[0])
        turn = self.snakeAI.selectDirection(spaces, state[1])
        dir = (self.direction + turn + 3) % 4
        return dir

    def isDead(self, gridWidth, gridHeight):
        head = self.spaces[0]
        if head.x < 0 or head.x >= gridWidth or head.y < 0 or head.y >= gridHeight:
            return True
        for space in self.spaces[1:]:
            if(space.equals(head)):
                return True
        return False

    def saveData(self, data):
        self.snakeNet.saveData(data)
    
    def hitFood(self, food):
        return food.equals(self.spaces[0])
    
    def fat(self, amt):
        self.grow += amt